esr/esr_test.go

```diff
@@ -0,0 +1,614 @@
+				r = httptest.NewRequest(http.MethodGet, "http://localhost/reg", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+
+				return w, r
+			},
+			"Good case GET, everything passes",
+		},
+		{
+			http.StatusBadRequest,
+			func(ua *UnitAsset) (w *httptest.ResponseRecorder, r *http.Request) {
+				ua.leading = true
+
+				w = httptest.NewRecorder()
+				body := io.NopCloser(strings.NewReader("{}"))
+				r = httptest.NewRequest(http.MethodPost, "http://localhost/reg", body)
+				r.Header = map[string][]string{}
+
+				return w, r
+			},
+			"Bad case POST, can't parse Content-Type from header",
+		},
+		{
+			http.StatusBadRequest,
+			func(ua *UnitAsset) (w *httptest.ResponseRecorder, r *http.Request) {
+				ua.leading = true
+
+				w = httptest.NewRecorder()
+				body := io.NopCloser(errReader(0))
+				r = httptest.NewRequest(http.MethodPost, "http://localhost/reg", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+
+				return w, r
+			},
+			"Bad case POST, error while reading body",
+		},
+		{
+			http.StatusBadRequest,
+			func(ua *UnitAsset) (w *httptest.ResponseRecorder, r *http.Request) {
+				ua.leading = true
+
+				w = httptest.NewRecorder()
+				body := io.NopCloser(strings.NewReader("{}"))
+				r = httptest.NewRequest(http.MethodPost, "http://localhost/reg", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+
+				return w, r
+			},
+			"Bad case POST, error while unpacking body",
+		},
+		{
+			http.StatusInternalServerError,
+			func(ua *UnitAsset) (w *httptest.ResponseRecorder, r *http.Request) {
+				ua.leading = true
+
+				w = httptest.NewRecorder()
+				body := io.NopCloser(strings.NewReader(`{"id": 0, "version":"SignalA_v1.0"}`))
+				r = httptest.NewRequest(http.MethodPost, "http://localhost/reg", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+
+				return w, r
+			},
+			"Bad case POST, request returns error",
+		},
+		{
+			http.StatusOK,
+			func(ua *UnitAsset) (w *httptest.ResponseRecorder, r *http.Request) {
+				ua.leading = true
+
+				w = httptest.NewRecorder()
+				body := io.NopCloser(strings.NewReader(`{"id": 0, "version":"ServiceQuest_v1"}`))
+				r = httptest.NewRequest(http.MethodPost, "http://localhost/reg", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+
+				return w, r
+			},
+			"Good case POST, request returns a result",
+		},
+		{
+			http.StatusMethodNotAllowed,
+			func(ua *UnitAsset) (w *httptest.ResponseRecorder, r *http.Request) {
+				ua.leading = true
+
+				w = httptest.NewRecorder()
+				body := io.NopCloser(strings.NewReader(`{"id": 0, "version":"ServiceQuest_v1"}`))
+				r = httptest.NewRequest(http.MethodDelete, "http://localhost/reg", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+
+				return w, r
+			},
+			"Bad case default, unsupported http method",
+		},
+	}
```

Comment:
Same here, don't abuse the setup functions. If they are doing the same thing anyway, might as well put it in the test function directly.